// gpt-tokenizer/cjs/specialTokens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HarmonyCall = exports.HarmonyConstrain = exports.HarmonyReturn = exports.HarmonyChannel = exports.HarmonyMessage = exports.HarmonyEnd = exports.HarmonyStart = exports.HarmonyStartOfText = exports.EndOfPrompt = exports.ImSep = exports.ImEnd = exports.ImStart = exports.FimSuffix = exports.FimMiddle = exports.FimPrefix = exports.EndOfText = void 0;
exports.EndOfText = '<|endoftext|>';
exports.FimPrefix = '<|fim_prefix|>';
exports.FimMiddle = '<|fim_middle|>';
exports.FimSuffix = '<|fim_suffix|>';
exports.ImStart = '<|im_start|>'; // 100264
exports.ImEnd = '<|im_end|>'; // 100265
exports.ImSep = '<|im_sep|>'; // 100266
exports.EndOfPrompt = '<|endofprompt|>';
exports.HarmonyStartOfText = '<|startoftext|>';
exports.HarmonyStart = '<|start|>';
exports.HarmonyEnd = '<|end|>';
exports.HarmonyMessage = '<|message|>';
exports.HarmonyChannel = '<|channel|>';
exports.HarmonyReturn = '<|return|>';
exports.HarmonyConstrain = '<|constrain|>';
exports.HarmonyCall = '<|call|>';
